lib/datadog/core/remote/configuration/repository.rb

```diff
@@ -67,19 +68,39 @@ class State
             attr_reader \
               :root_version,
               :targets_version,
-              :config_states,
               :has_error,
               :error,
-              :opaque_backend_state
+              :opaque_backend_state,
+              :repository
 
             def initialize(repository)
+              @repository = repository
               @root_version = repository.root_version
               @targets_version = repository.targets_version
               @config_states = []
               @has_error = false
               @error = ''
               @opaque_backend_state = repository.opaque_backend_state
             end
+
+            def config_states
+              return [] if @repository.contents.empty?
+
+              @repository.contents.each_with_object([]) do |content, acc|
+                config_state = {
+                  path: content.path.to_s,
+                  length: content.length,
+                  hashes: content.hashes.each_with_object([]) do |(key, value), hashes_acc|
+                    hashes_acc << {
+                      algorithm: key,
+                      hash: value
+                    }
+                    hashes_acc
+                  end
```

Comment:
- `|(k, v), a|` is a well known pattern, no need to spell them out loud, that is, except for actually giving them names

  ```ruby
  hashes: content.hashes.each_with_object([]) do |(type, hexdigest), hashes|
    {
      algorithm: type,
      hash: hexdigest
    }
    hashes
  end
  ```

- `each_with_object` provides an object to mutate, so contrary to `reduce` there is no need to return it from the block

  ```ruby
  hashes: content.hashes.each_with_object([]) do |(type, hexdigest), hashes|
    {
      algorithm: type,
      hash: hexdigest
    }
  end
  ```

- even if it was so, `<<` returns the new array value
- since there is no skipping, this s actually `map`

  ```ruby
  hashes: content.hashes.map do |type, hexdigest|
    {
      algorithm: type,
      hash: hexdigest
    }
  end
  ```
  
- which we can turn into one line:

  ```ruby
  hashes: content.hashes.map { |type, hexdigest| { algorithm: type, hash: hexdigest } }
  ```

- and then, following Andrew Gerrand's rule of distance for names†, optionally: 

  ```ruby
  hashes: content.hashes.map { |t, h| { algorithm: t, hash: h } }
  ```

† [What's in a name?](https://youtu.be/sFUSP8Au_PE?t=223)

Reply:
The same process can be applied to the parent block and its leading line:

```
@repository.contents.each_with_object([]) do |content, acc|
```

Reply:
Great detailed explanation @lloeki Thanks 😄 

I improved the code following your feedback